modules/light-clients/08-wasm/types/update_test.go

```diff
@@ -652,115 +662,117 @@ func (suite *TypesTestSuite) TestUpdateStateGrandpa() {
 
 	testCases := []struct {
-		name    string
-		setup   func()
-		expPass bool
+		name               string
+		malleate           func()
+		panicErr           error
+		updatedClientState []byte
 	}{
 		{
-			"successful update",
+			"success: no update",
 			func() {
-				data, err := base64.StdEncoding.DecodeString(suite.testData["misbehaviour"])
-				suite.Require().NoError(err)
-				clientMsg = &types.ClientMessage{
-					Data: data,
-				}
-
-				clientState, ok = suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientState(suite.ctx, defaultWasmClientID)
-				suite.Require().True(ok)
+				suite.mockVM.RegisterSudoCallback(types.UpdateStateOnMisbehaviourMsg{}, func(_ wasmvm.Checksum, _ wasmvmtypes.Env, sudoMsg []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					var msg *types.SudoMsg
+
+					err := json.Unmarshal(sudoMsg, &msg)
+					suite.Require().NoError(err)
+
+					suite.Require().NotNil(msg.UpdateStateOnMisbehaviour)
+					suite.Require().NotNil(msg.UpdateStateOnMisbehaviour.ClientMessage)
+					suite.Require().Nil(msg.UpdateState)
+					suite.Require().Nil(msg.UpdateState)
+					suite.Require().Nil(msg.VerifyMembership)
+					suite.Require().Nil(msg.VerifyNonMembership)
+					suite.Require().Nil(msg.VerifyUpgradeAndUpdateState)
+					suite.Require().Nil(msg.CheckSubstituteAndUpdateState)
+
+					resp, err := json.Marshal(types.EmptyResult{})
+					if err != nil {
+						return nil, 0, err
+					}
+
+					return &wasmvmtypes.Response{
+						Data: resp,
+					}, types.DefaultGasUsed, nil
+				})
 			},
-			true,
+			nil,
+			nil,
+		},
+		{
+			"success: client state updated on valid misbehaviour",
+			func() {
+				suite.mockVM.RegisterSudoCallback(types.UpdateStateOnMisbehaviourMsg{}, func(_ wasmvm.Checksum, _ wasmvmtypes.Env, sudoMsg []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					var msg *types.SudoMsg
+					err := json.Unmarshal(sudoMsg, &msg)
+					suite.Require().NoError(err)
+
+					// set new client state in store
+					store.Set(host.ClientStateKey(), msg.UpdateStateOnMisbehaviour.ClientMessage.Data)
+					resp, err := json.Marshal(types.EmptyResult{})
+					if err != nil {
+						return nil, 0, err
+					}
+
+					return &wasmvmtypes.Response{Data: resp}, types.DefaultGasUsed, nil
+				})
+			},
+			nil,
+			types.MockClientStateBz,
+		},
+		{
+			"failure: invalid client message",
+			func() {
+				clientMsg = &ibctmtypes.Header{}
+				suite.mockVM.RegisterSudoCallback(types.UpdateStateOnMisbehaviourMsg{}, func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					// empty resp because this test case does not reach the VM
```

Comment:
do we even need to register a callback here if its unreachable?

Reply:
true, good call, i can remove this!